client/asset/dcr/dcr.go

```diff
@@ -2157,85 +2180,40 @@ func (dcr *ExchangeWallet) refundTx(coinID, contract dex.Bytes, val uint64, refu
 
 // Address returns an address for the exchange wallet.
 func (dcr *ExchangeWallet) Address() (string, error) {
-	addr, err := dcr.wallet.GetNewAddressGapPolicy(dcr.ctx, dcr.acct, dcrwallet.GapPolicyIgnore)
+	addr, err := dcr.wallet.ExternalAddress(dcr.ctx)
 	if err != nil {
 		return "", err
 	}
 	return addr.String(), nil
 }
 
-func (dcr *ExchangeWallet) accountUnlocked(ctx context.Context, acct string) (encrypted, unlocked bool, err error) {
-	var res *walletjson.AccountUnlockedResult
-	res, err = dcr.wallet.AccountUnlocked(ctx, acct)
-	if err != nil {
-		return
-	}
-	encrypted = res.Encrypted
-	if res.Unlocked != nil { // should only be when encrypted
-		unlocked = *res.Unlocked
-	}
-	return
-}
-
 // Unlock unlocks the exchange wallet.
 func (dcr *ExchangeWallet) Unlock(pw []byte) error {
-	encryptedAcct, unlocked, err := dcr.accountUnlocked(dcr.ctx, dcr.acct)
+	unlocked, err := dcr.wallet.Unlocked(dcr.ctx)
 	if err != nil {
 		return err
 	}
-	if !encryptedAcct {
-		return dcr.wallet.UnlockWallet(dcr.ctx, string(pw), 0)
-
-	}
 	if unlocked {
 		return nil
 	}
 
-	return dcr.wallet.UnlockAccount(dcr.ctx, dcr.acct, string(pw))
+	return dcr.wallet.Unlock(dcr.ctx, pw)
 }
 
 // Lock locks the exchange wallet.
 func (dcr *ExchangeWallet) Lock() error {
-	if dcr.wallet.Disconnected() {
-		return asset.ErrConnectionDown
-	}
-
-	// Since hung calls to Lock() may block shutdown of the consumer and thus
-	// cancellation of the ExchangeWallet subsystem's Context, dcr.ctx, give
-	// this a timeout in case the connection goes down or the RPC hangs for
-	// other reasons.
-	ctx, cancel := context.WithTimeout(dcr.ctx, 5*time.Second)
-	defer cancel()
-
-	encryptedAcct, unlocked, err := dcr.accountUnlocked(ctx, dcr.acct)
-	if err != nil {
-		return err
-	}
-	if !encryptedAcct {
-		return dcr.wallet.LockWallet(ctx)
-	}
-	if !unlocked {
-		return nil
-	}
-
-	return dcr.wallet.LockAccount(dcr.ctx, dcr.acct)
+	return dcr.wallet.Lock(dcr.ctx)
 }
 
 // Locked will be true if the wallet is currently locked.
 // Q: why are we ignoring RPC errors in this?
 func (dcr *ExchangeWallet) Locked() bool {
-	// First return locked status of the account, falling back to walletinfo if
-	// the account is not individually password protected.
-	encrypted, unlocked, err := dcr.accountUnlocked(dcr.ctx, dcr.acct)
+	unlocked, err := dcr.wallet.Unlocked(dcr.ctx)
```

Comment:
This makes it the responsibility of the `Wallet` impl to check if the account is encrypted and perform an account unlocking rather than a wallet unlocking but the method says `wallet.Unlock`. At the very least, the doc for `wallet.Unlock` should mention this. But I think it's better to keep the account and wallet lock/unlock/locked methods separate. The `Wallet` interface does not represent an `ExchangeWallet` but a real dcr wallet, the semantics should resemble that of a real dcr wallet while the `ExchangeWallet` makes use of the actual wallet methods as needed, in this case, checking account lock status before trying to perform wallet unlocking.

Reply:
Account vs wallet unlocking is implemented by dcrwallet and is a new feature. I definitely wouldn't assume that all wallets will differentiate. Regardless, this is the most general form, so should be suitable for any implementation. What happens under the hood is up to the `Wallet`.

Reply:
It's a good point though, and I'll mention something in the docs about account vs. wallet unlocking.